src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -174,7 +174,77 @@ public bool IsValid
+        /// errors has already been recorded, ensures that a <see cref="TooManyModelErrorsException"/> exception is
+        /// recorded instead.
+        /// </summary>
+        /// <remarks>
+        /// This overload allows adding the <paramref name="exception"/> to the current <see cref="ModelStateDictionary"/>
+        /// when <see cref="ModelMetadata"/> is not available and/or the <paramref name="exception"/> has custom
+        /// or relevant information regarding the validation state.
+        /// For cases in which <see cref="ModelMetadata"/> is available, it is recommended that the overload
+        /// with <see cref="ModelMetadata"/> as a parameter be used instead.
+        /// </remarks>
+        /// <param name="key">The key of the <see cref="ModelStateEntry"/> to add errors to.</param>
+        /// <param name="exception">The <see cref="Exception"/> to add.</param>
+        public void AddModelError(string key, Exception exception)
+        {
+            TryAddModelError(key, exception);
+        }
+
+        /// <summary>
+        /// Attempts to add the specified <paramref name="exception"/> to the <see cref="ModelStateEntry.Errors"/>
+        /// instance that is associated with the specified <paramref name="key"/>. If the maximum number of allowed
+        /// errors has already been recorded, ensures that a <see cref="TooManyModelErrorsException"/> exception is
+        /// recorded instead.
+        /// </summary>
+        /// <remarks>
+        /// This overload allows adding the <paramref name="exception"/> to the current <see cref="ModelStateDictionary"/>
+        /// when <see cref="ModelMetadata"/> is not available and/or the <paramref name="exception"/> has custom
+        /// or relevant information regarding the validation state.
+        /// For cases in which <see cref="ModelMetadata"/> is available, it is recommended that the overload
+        /// with <see cref="ModelMetadata"/> as a parameter be used instead.
+        /// </remarks>
+        /// <param name="key">The key of the <see cref="ModelStateEntry"/> to add errors to.</param>
+        /// <param name="exception">The <see cref="Exception"/> to add.</param>
+        /// <returns>
+        /// <c>True</c> if the given error was added, <c>false</c> if the error was ignored.
+        /// See <see cref="MaxAllowedErrors"/>.
+        /// </returns>
+        public bool TryAddModelError(string key, Exception exception)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (exception == null)
+            {
+                throw new ArgumentNullException(nameof(exception));
+            }
+
+            if (ErrorCount >= MaxAllowedErrors - 1)
+            {
+                EnsureMaxErrorsReachedRecorded();
+                return false;
+            }
+
+            if (exception is InputFormatterException && !string.IsNullOrEmpty(exception.Message))
```

Comment:
Why does this block exist? Makes the method behave rather differently from its description, especially since one of its documented purposes is to maintain the `exception` for later use (regardless of its type). Might need to check with others but I'm not a fan of including this special case here. (I'd feel differently if we had any reason to use this overload within MVC.)

Reply:
My guess was that this exception type is typically thrown from our code only so it's safe to only expose the exception message. But this overload probably doesn't need it for that reason. Removing it. cc @Eilon

Reply:
I think it's important to keep all the methods behaving the same, which is to special-case this type. Otherwise if our own framework calls this new method (which it certainly could) it would behave incorrectly.

Reply:
@Eilon it would always be a bug if our own code calls an overload that lacks the `ModelMetadata` parameter, `FormatException` and `OverflowException` would not be handled correctly.

That said, if we include this `InputFormatterException` special case, we need to get rid of the words about preserving the `<paramref name="exception"/>`. IOW these overloads would just be for those who lack `ModelMetadata`, not really those who want to avoid our special handling of some `Exception` types.

Reply:
@Eilon this method is already different because it DOES NOT special case `FormatException` or `OverflowException` either. Is that something that needs to happen too? Doing that will not be trivial without `ModelMetadata` https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs#L245.